### ISSUE TITLE
Errors installing on macOS Sierra

### ISSUE THREAD

**Issue by cudmore** [LeeKamentsky/python-javabridge#159]
Earlier today I successfully installed javabridge on macOS Mojave. Now I am trying to install on macOS Sierra (10.12.6) and getting java version 6 errors?

When I issue

```
pip install javabridge
```

I get these errors

```
Installing collected packages: javabridge
  Running setup.py install for javabridge ... error
    Complete output from command /Users/cudmore/Dropbox/olympus/myBioformats/bio_env/bin/python3 -u -c "import setuptools, tokenize;__file__='/private/var/folders/zy/gh9wp3pd2_d4hhllv2hzt5gh0000gn/T/pip-install-dy6a0ueu/javabridge/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /private/var/folders/zy/gh9wp3pd2_d4hhllv2hzt5gh0000gn/T/pip-record-lcye5bz0/install-record.txt --single-version-externally-managed --compile --install-headers /Users/cudmore/Dropbox/olympus/myBioformats/bio_env/include/site/python3.7/javabridge:
    running install
    running build
    running build_py
    creating build
    creating build/lib.macosx-10.12-x86_64-3.7
    creating build/lib.macosx-10.12-x86_64-3.7/javabridge
    copying javabridge/__init__.py -> build/lib.macosx-10.12-x86_64-3.7/javabridge
    copying javabridge/_version.py -> build/lib.macosx-10.12-x86_64-3.7/javabridge
    copying javabridge/jutil.py -> build/lib.macosx-10.12-x86_64-3.7/javabridge
    copying javabridge/locate.py -> build/lib.macosx-10.12-x86_64-3.7/javabridge
    copying javabridge/noseplugin.py -> build/lib.macosx-10.12-x86_64-3.7/javabridge
    copying javabridge/wrappers.py -> build/lib.macosx-10.12-x86_64-3.7/javabridge
    creating build/lib.macosx-10.12-x86_64-3.7/javabridge/tests
    copying javabridge/tests/__init__.py -> build/lib.macosx-10.12-x86_64-3.7/javabridge/tests
    copying javabridge/tests/test_cpython.py -> build/lib.macosx-10.12-x86_64-3.7/javabridge/tests
    copying javabridge/tests/test_javabridge.py -> build/lib.macosx-10.12-x86_64-3.7/javabridge/tests
    copying javabridge/tests/test_jutil.py -> build/lib.macosx-10.12-x86_64-3.7/javabridge/tests
    copying javabridge/tests/test_wrappers.py -> build/lib.macosx-10.12-x86_64-3.7/javabridge/tests
    creating build/lib.macosx-10.12-x86_64-3.7/javabridge/jars
    copying javabridge/jars/cpython.jar -> build/lib.macosx-10.12-x86_64-3.7/javabridge/jars
    copying javabridge/jars/rhino-1.7R4.jar -> build/lib.macosx-10.12-x86_64-3.7/javabridge/jars
    copying javabridge/jars/runnablequeue.jar -> build/lib.macosx-10.12-x86_64-3.7/javabridge/jars
    copying javabridge/jars/test.jar -> build/lib.macosx-10.12-x86_64-3.7/javabridge/jars
    running build_ext
    javac -source 6 -target 6 /private/var/folders/zy/gh9wp3pd2_d4hhllv2hzt5gh0000gn/T/pip-install-dy6a0ueu/javabridge/java/org/cellprofiler/runnablequeue/RunnableQueue.java
    warning: [options] bootstrap class path not set in conjunction with -source 6
    error: Source option 6 is no longer supported. Use 7 or later.
    error: Target option 6 is no longer supported. Use 7 or later.
    error: command 'javac' failed with exit status 2
    
    ----------------------------------------
Command "/Users/cudmore/Dropbox/olympus/myBioformats/bio_env/bin/python3 -u -c "import setuptools, tokenize;__file__='/private/var/folders/zy/gh9wp3pd2_d4hhllv2hzt5gh0000gn/T/pip-install-dy6a0ueu/javabridge/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /private/var/folders/zy/gh9wp3pd2_d4hhllv2hzt5gh0000gn/T/pip-record-lcye5bz0/install-record.txt --single-version-externally-managed --compile --install-headers /Users/cudmore/Dropbox/olympus/myBioformats/bio_env/include/site/python3.7/javabridge" failed with error code 1 in /private/var/folders/zy/gh9wp3pd2_d4hhllv2hzt5gh0000gn/T/pip-install-dy6a0ueu/javabridge/
```

'/usr/libexec/java_home' reports

```
/Library/Java/JavaVirtualMachines/jdk-12.0.1.jdk/Contents/Home
```

'java -version' reports

```
java version "12.0.1" 2019-04-16
Java(TM) SE Runtime Environment (build 12.0.1+12)
Java HotSpot(TM) 64-Bit Server VM (build 12.0.1+12, mixed mode, sharing)
```

'javac -version' reports:

```
javac 12.0.1
```

I created a 'JAVA_HOME' environment variable with:

```
echo export "JAVA_HOME=\$(/usr/libexec/java_home)" >> ~/.bash_profile
```
'echo $JAVA_HOME' reports:

```
/Library/Java/JavaVirtualMachines/jdk-12.0.1.jdk/Contents/Home
```

I tried 'pip install --no-cache-dir javabridge' and got the same errors.

Now I am entering unfamiliar territory. Java does not seem to like the installer using java 6? I then cloned the repository and tried installing from source after changing 'javabridge-1.0.18/setup.py'

```
        # was this
        #javac_command = [javac_loc, "-source", "6", "-target", "6"]
        # now this
        javac_command = [javac_loc, "-source", "8", "-target", "8"]
```

Then running 'python setup.py install' yields:

```
running install
running bdist_egg
running egg_info
writing javabridge.egg-info/PKG-INFO
writing dependency_links to javabridge.egg-info/dependency_links.txt
writing entry points to javabridge.egg-info/entry_points.txt
writing requirements to javabridge.egg-info/requires.txt
writing top-level names to javabridge.egg-info/top_level.txt
reading manifest file 'javabridge.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
no previously-included directories found matching '*.pyx'
writing manifest file 'javabridge.egg-info/SOURCES.txt'
installing library code to build/bdist.macosx-10.12-x86_64/egg
running install_lib
running build_py
running build_ext
javac -source 8 -target 8 java/org/cellprofiler/runnablequeue/RunnableQueue.java
warning: [options] bootstrap class path not set in conjunction with -source 8
1 warning
javac -source 8 -target 8 java/org/cellprofiler/javabridge/test/RealRect.java
warning: [options] bootstrap class path not set in conjunction with -source 8
1 warning
javac -source 8 -target 8 java/org/cellprofiler/javabridge/CPython.java java/org/cellprofiler/javabridge/CPythonInvocationHandler.java
warning: [options] bootstrap class path not set in conjunction with -source 8
Note: java/org/cellprofiler/javabridge/CPythonInvocationHandler.java uses unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
1 warning
building 'javabridge._javabridge' extension
clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Library/Java/JavaVirtualMachines/jdk-12.0.1.jdk/Contents/Home/include -I/Library/Java/JavaVirtualMachines/jdk-12.0.1.jdk/Contents/Home/include/darwin -I/usr/local/include -I/usr/local/opt/openssl/include -I/usr/local/opt/sqlite/include -I/Users/cudmore/Dropbox/olympus/myBioformats/bio_env/include -I/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/include/python3.7m -c _javabridge.c -o build/temp.macosx-10.12-x86_64-3.7/_javabridge.o
_javabridge.c:525:10: fatal error: 'numpy/arrayobject.h' file not found
#include "numpy/arrayobject.h"
         ^~~~~~~~~~~~~~~~~~~~~
1 error generated.
error: command 'clang' failed with exit status 1
```

Bottom line, how do I install javabridge on macOS Sierra???

Please help.

**Comment by LeeKamentsky** [comment]
Hi Robert,
I think the key here is `-I/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/include/python3.7m`. and `fatal error: 'numpy/arrayobject.h' file not found`. The proximate cause is that the compiler can't find one of the Numpy C include files it needs. There are a couple of possibilities:
* Numpy not installed. You can check this by
```python -c "import numpy;print(numpy.__file__)"```
It should give you the location for Numpy, e.g. `/usr/local/Cellar/3.7.0/Frameworks/Python.framework/Versions/3.7... something or other`
* Numpy installed in the wrong place - it could be that Python is picking up Numpy from one place and the compiler is looking for the Numpy includes in another. The numpy.__file__ location might give us a clue there.
* Numpy mis-installed - perhaps something has happened to the include files. You can check the path, `/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/include/python3.7m/numpy/arrayobject.h` and the directories above that and see if it looks like something is missing.

I have to say,  I don't have access to a Mac and Linux and Windows are my strengths, not the Mac, so beyond the above, I may not be much help, especially if it's something wrong with the way things build on Sierra.
--Lee

**Comment by cudmore** [comment]
Thanks for your troubleshooting. Yes, it was because the compiler could not find the numpy headers. I was able to fix it following the suggestion in [this blog post](http://notesbyanerd.com/2018/11/02/pip-install-cant-find-numpy-header/).

In python:

```
import numpy as np
np.get_include()
```

np.get_include() printed '/usr/local/lib/python3.7/site-packages/numpy/core/include'.

Then, at a command prompt:

```
export CFLAGS="-I/usr/local/lib/python3.7/site-packages/numpy/core/include"
```

And then 'pip install javabridge' was successful and I was on my way using javabridge! Thanks for your help and hope this helps others.
